packages/plugin-ext/src/plugin/languages/selection-range.ts

```diff
@@ -0,0 +1,72 @@
+/********************************************************************************
+ * Copyright (C) 2020 Red Hat, Inc. and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import * as theia from '@theia/plugin';
+import { DocumentsExtImpl } from '../documents';
+import { URI } from 'vscode-uri/lib/umd';
+import * as model from '../../common/plugin-api-rpc-model';
+import * as Converter from '../type-converters';
+import * as types from '../types-impl';
+
+ export class SelectionRangeProviderAdapter {
```

Comment:
Is it copied from MS codebase?

Reply:
Yep, that's why it is in draft. It copied from [here](https://github.com/microsoft/vscode/blob/standalone/0.19.x/src/vs/workbench/api/common/extHostLanguageFeatures.ts#L1107-L1151) and I'm not sure under what license it should be. The same as [here](https://github.com/microsoft/vscode/blob/standalone/0.19.x/src/vs/workbench/api/common/extHostLanguageFeatures.ts#L1-L4)?

Reply:
Yes, please do the same as for other adapters.

Reply:
Do you need a help with CQ?

Reply:
>Do you need help with a CQ?

Once, I created a CQ for third party library, I'll try to figure it out. With a CQ for a single file there is not big difference? Include [extHostLanguageFeatures.ts](https://github.com/microsoft/vscode/blob/standalone/0.19.x/src/vs/workbench/api/common/extHostLanguageFeatures.ts) and [selection-range.ts](https://github.com/eclipse-theia/theia/pull/7534/files/eb5656eded4dce711440522a9d47ace8cb0f8490#diff-c8213b8343ad9c25314f311d40ded5de)(with an original license from vscode) into request?

Reply:
You should follow: https://github.com/eclipse-theia/theia/wiki/Registering-CQs#case-3rd-party-project-code-copiedforked-from-another-project-into-eclipse-theia-maintained-by-us yes, there should no be much difference

It is better to do it now, getting the CQ through sometimes is a long process.